setuptools/glob.py

```diff
@@ -129,24 +149,24 @@ def _rlistdir(dirname):
         return
     for x in names:
         yield x
-        path = os.path.join(dirname, x) if dirname else x
+        # mypy false-positives: str or bytes type possibility is always kept in sync
+        path = os.path.join(dirname, x) if dirname else x  # type: ignore[arg-type]
         for y in _rlistdir(path):
-            yield os.path.join(x, y)
+            yield os.path.join(x, y)  # type: ignore[arg-type]
 
 
 magic_check = re.compile('([*?[])')
 magic_check_bytes = re.compile(b'([*?[])')
 
 
-def has_magic(s):
+def has_magic(s: str | bytes) -> bool:
     if isinstance(s, bytes):
-        match = magic_check_bytes.search(s)
+        return magic_check_bytes.search(s) is not None
     else:
-        match = magic_check.search(s)
-    return match is not None
+        return magic_check.search(s) is not None
```

Comment:
mypy will declare the first `match` as typed `Math[bytes] | None`, so the second assignement doesn't respect that typing, this is the same issue as https://github.com/python/mypy/issues/10736 AFAICT

Either the variable needs to be explicitely typed as `Match[bytes] | Match[str] | None`, or we simply don't assign.